### PR TITLE
[SES5] qa: introducing the deepsea.repository subtask and automated upgrade testing

### PR DESCRIPTION
**Pull request by smithfarm**
TODO:
- [x] figure out how to distinguish between a Cluster object and a Remote object
- [x] `zypper rm` conflicting RPMs on all nodes, not just on master
- [x] enforce deepsea must be installed from RPM
- [x] `--suite deepsea/tier4/upgrade` gets all the way to `ceph.maintenance.upgrade`

### CODE REVIEW

File: qa/suites/deepsea/tier4/upgrade/3-test-phase.yaml [hunk 1]
```diff
@@ -1,9 +1,37 @@
...
+                        - name: ses
+                          url: http://10.86.0.120/artifacts/ci/d4ad635e5108134c42c5e5e477a4224a320ca2a57e0795865712f7f2f7ebf1ec/SUSE-Enterprise-Storage-6-POOL-x86_64-Build5.6
+                        - name: basesystem
+                          url: http://10.86.0.120/artifacts/ci/e9bf06179ab8222426cd8b22e7e5d63f40d2b12f67d6a47a4898983f5735ccdc/SLE-15-Module-Basesystem-POOL-x86_64-Build668.1
+                        - name: server
+                          url: http://10.86.0.120/artifacts/ci/618743a8bcd28a64b2881bd11eaccb24f17fe887576e109f59a64e46f9564f0a/SLE-15-Module-Server-Applications-POOL-x86_64-Build668.1
+                        - name: internal
+                          url: http://10.86.0.120/artifacts/ci/55d6cc29675bc77b0b44bd3063b830547a1f271e8e42c96e4729acf1dadb0209/SUSE-Enterprise-Storage-6-POOL-Internal-x86_64-Build5.2
+                        - name: update
+                          url: http://10.86.0.120/artifacts/ci/snapshot/201811060915/3577b41f35e5000f7bd8aef985ed8b32cbeb766d07cb7efbdbb8234e935d3789/SLE-15-Update
```
Comment (smithfarm) [inline qa/suites/deepsea/tier4/upgrade/3-test-phase.yaml]: this bit of YAML will be dropped - the person running the test (or Jenkins) will have to provide this bit of yaml via the command line, with the URLs of the latest milestone/version of SES6

File: qa/tasks/scripts.py [hunk 2]
```diff
@@ -19,8 +19,17 @@ def __init__(self, ctx, logger):
             ctx['scripts_copied'] = True
 
     def run(self, remote, script_name, args=[], as_root=True):
-        self.log.info('(scripts) running script {} on remote {}'
-                      .format(script_name, remote.hostname)
+        remote_type = type(remote)
+        self.log.debug(
+            '(scripts) run method was passed a remote object of type {}'
+            .format(remote_type)
+            )
```
Comment (smithfarm) [inline qa/tasks/scripts.py]: in the test log, this comes out as `(scripts) run method was passed a remote object of type <class 'teuthology.orchestra.cluster.Cluster'>`

File: qa/tasks/scripts.py [hunk 3]
```diff
@@ -19,8 +19,17 @@ def __init__(self, ctx, logger):
...
+        if remote_type == 'Cluster':
+            remote_spec = 'the whole cluster'
+        else:
+            remote_spec = 'remote {}'.format(remote.hostname)
```
Comment (smithfarm) [inline qa/tasks/scripts.py]: and then this conditional fails to recognize that this object is of class Cluster and the job fails lamely with `AttributeError: 'Cluster' object has no attribute 'hostname'`

(@smithfarm needs to learn object introspection in Python properly.)

**Comment by smithfarm** [comment]
@jschmid1 The latest test run gets all the way to `ceph.maintenance.upgrade` and fails with some (for me) obscure Salt errors: http://10.86.0.135/ubuntu-2018-11-30_09:00:50-deepsea:tier4:upgrade-wip-qa-repos---basic-openstack/943/teuthology.log

Do you think those errors are related to this PR?

**Comment by jschmid1** [comment]
In the logs I also saw this:

```
File "/var/cache/salt/minion/extmods/modules/retry.py", line 29
print line
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(line)?
```
also the case for the rgw.py module



> Do you think those errors are related to this PR?

No I don't think so. I think they are genuine issues with the upgrade and deepsea

**Comment by jschmid1** [comment]
Looking in `deepsea-master`, I can't find those py2-style print statements, which indicates that the modules weren't synced properly

**Comment by jschmid1** [comment]
lgtm now!

**Pull request merged**